tensorboard/loader.py

```diff
@@ -0,0 +1,414 @@
+      with self._lock:
+        if size > self._size:
+          self._size = size
+    return size
+
+  def get_next_record(self):
+    """Reads one record.
+
+    When this method is first called, it will spawn the thread and
+    block until a record is read. Once the thread starts, it will queue
+    up records which can be read without blocking. The exception is
+    when we reach the end of the file, in which case each repeated call
+    will be synchronous. There is no background polling. If new data is
+    appended to the file, new records won't be buffered until this
+    method is invoked again. The caller should take care to meter calls
+    to this method once it reaches the end of file, lest they impact
+    performance.
+
+    Returns:
+      A Record object, or None if there are no more records available
+      at the moment.
+
+    Raises:
+      IOError: If this instance has been closed.
+      Exception: To propagate any exceptions that may have been thrown
+          by the read operation in the other thread. If an exception is
+          thrown, then all subsequent calls to this method will rethrow
+          that same exception.
+
+    :rtype: Record
+    """
+    with self._lock:
+      if self._is_closed:
+        raise IOError('%s is closed' % self)
+      if not self._is_started:
+        self._thread.start()
+        self._is_started = True
+      else:
+        record = self._get_record()
+        if record is not None:
+          return record
+        self._is_tapped = False
+        self._please_sir_i_want_some_more.notify()
+      while (self._read_exception is None and
+             not self._is_tapped and
+             not self._records):
+        self._no_soup_for_you.wait()
+      return self._get_record()
+
+  def close(self):
+    """Closes event log reader if open.
+
+    If the i/o thread is running, this method blocks until it has been
+    shut down.
+
+    Further reads are not permitted after this method is called.
+
+    Raises:
+      Exception: To propagate any exceptions that may have been thrown
+          by the close operation in the other thread. If an exception
+          is thrown, then all subsequent calls to this method will
+          rethrow that same exception.
+    """
+    with self._lock:
+      if not self._is_closed:
+        self._is_closed = True
+        if not self._is_started:
+          return
+        self._please_sir_i_want_some_more.notify()
+      while self._reader is not None:
+        self._no_soup_for_you.wait()
+      if self._close_exception is not None:
+        raise self._close_exception
+
+  def _get_record(self):
+    if self._read_exception is not None:
+      raise self._read_exception
+    if not self._records:
+      return None
+    record = self._records.popleft()
+    self._buffered -= len(record.record)
+    if self._should_wakeup():
+      self._please_sir_i_want_some_more.notify()
+    return record
+
+  def _run(self):
```

Comment:
one idea that I'd like having the concept of a generalized async ReadaheadBufferer loader being factored out into an abstract class, and then making the BufferedRecordReader the composition of the ReadaheadBufferer and the implementation details specific to the RecordReader. it would make the contract and testing more clear. is this feasible?